include/boost/random/niederreiter_base2.hpp

```diff
@@ -0,0 +1,391 @@
+    for (std::size_t dim = 0; dim != dimension; ++dim)
+    {
+      const int poly = nb2::niederreiter_tables::polynomial(dim);
+      if (static_cast<std::size_t>(poly) > 
+          static_cast<std::size_t>(std::numeric_limits<IntType>::max())) {
+        boost::throw_exception( std::range_error("niederreiter_base2: polynomial value outside the given IntType range") );
+      }
+
+      const int degree = multiprecision::detail::find_msb(poly); // integer log2(poly)
+      const int max_degree = degree * ((bit_count / degree) + 1);
+
+      v.resize(degree + max_degree);
+  
+      // For each dimension, we need to calculate powers of an
+      // appropriate irreducible polynomial, see Niederreiter
+      // page 65, just below equation (19).
+      // Copy the appropriate irreducible polynomial into PX,
+      // and its degree into E.  Set polynomial B = PX ** 0 = 1.
+      // M is the degree of B.  Subsequently B will hold higher
+      // powers of PX.
+      int pb_degree = 0;
+      boost::dynamic_bitset<> pb(max_degree, 1);
+  
+      int j = 0;
+      while (j < bit_count)
+      {
+        // Now multiply B by PX so B becomes PX**J.
+        // In section 2.3, the values of Bi are defined with a minus sign :
+        // don't forget this if you use them later!
+        nb2::modulo2_multiply(poly, boost::move(pb)).swap(pb);
+
+        // If U = 0, we need to set B to the next power of PX
+        // and recalculate V.
+        nb2::calculate_v(pb, pb_degree, v);
+  
+        // Niederreiter (page 56, after equation (7), defines two
+        // variables Q and U.  We do not need Q explicitly, but we
+        // do need U.
+  
+        // Advance Niederreiter's state variables.
+        for (int u = 0; u < degree && j < bit_count; ++u, ++j)
+        {
+          // Now C is obtained from V.  Niederreiter
+          // obtains A from V (page 65, near the bottom), and then gets
+          // C from A (page 56, equation (7)).  However this can be done
+          // in one step.  Here CI(J,R) corresponds to
+          // Niederreiter's C(I,J,R).
+          for (int r = 0; r < bit_count; ++r) {
+            ci[r][j] = v[r + u];
+          }
+        }
+      }
+  
+      // The array CI now holds the values of C(I,J,R) for this value
+      // of I.  We pack them into array CJ so that CJ(I,R) holds all
+      // the values of C(I,J,R) for J from 1 to NBITS.
+      for (int r = 0; r < bit_count; ++r)
+      {
+        IntType term = 0;
+        for (int j = 0; j < bit_count; ++j)
+          term = 2*term + ci[r][j];
+        bits[r][dim] = term;
+      }
+    }
+  }
+
+  value_type operator()(int i, int j) const
+  {
+    return bits[i][j];
+  }
+
+private:
+  boost::multi_array<IntType, 2> bits;
+};
+
+} // namespace detail
+/** @endcond */
+
+//!class template niederreiter_base2 implements a quasi-random number generator as described in
+//! \blockquote
+//!Bratley, Fox, Niederreiter, ACM Trans. Model. Comp. Sim. 2, 195 (1992).
```

Comment:
Shouldn't we deploy Neiderreiter's improved sequence from 1996 rather than the 1992 work?

https://www.sciencedirect.com/science/article/pii/S1071579796900167

Reply:
I will take a look.